lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+                    break
+
+            display_name = None
+            for user in users:
+                if user.display_name:
+                    display_name = user.display_name
+                    break
+
+            self._unified_users[h_userid] = UnifiedUser(
+                h_userid=h_userid,
+                users=tuple(users),
+                email=email,
+                display_name=display_name,
+            )
+
+        return self._unified_users
+
+    @property
+    def unified_courses(self):
+        """
+        Return a dict mapping authority_provided_id's to UnifiedCourse's.
+
+        Multiple keys in the dict may point to the same UnifiedCourse object,
+        meaning that unified_courses.values() may contain duplicates. This
+        happens when there are annotations in multiple sub-groupings that
+        belong to the same course.
+        """
+        if self._unified_courses is not None:
+            return self._unified_courses
+
+        self._unified_courses = {}
+
+        def is_instructor(unified_user: UnifiedUser, courses: Iterable[Course]) -> bool:
+            """Return True if `user` is an instructor in any of `courses`."""
+            return bool(
+                self._db.scalar(
+                    select(func.count())
+                    .select_from(AssignmentMembership)
+                    .join(Assignment)
+                    .join(AssignmentGrouping)
+                    .join(LTIRole)
+                    .filter(
+                        AssignmentGrouping.grouping_id.in_(
+                            (course.id for course in courses)
+                        )
+                    )
+                    .filter(
+                        AssignmentMembership.user_id.in_(
+                            user.id for user in unified_user.users
+                        )
+                    )
+                    .filter(LTIRole.type == "instructor", LTIRole.scope == "course")
+                )
+            )
+
+        authority_provided_ids = set(
+            annotation["group"]["authority_provided_id"]
+            for annotation in self._annotations
+        )
+
+        for authority_provided_id in authority_provided_ids:
+            if authority_provided_id in self._unified_courses:
+                continue
+
+            first_grouping = self._db.scalars(
+                select(Grouping)
+                .filter_by(authority_provided_id=authority_provided_id)
+                .order_by(Grouping.updated.desc())
+            ).first()
+
+            if not first_grouping:
+                continue
+
+            if first_grouping.type == Grouping.Type.COURSE:
+                course_authority_provided_id = first_grouping.authority_provided_id
+            else:
+                assert first_grouping.parent.type == Grouping.Type.COURSE
+                course_authority_provided_id = (
+                    first_grouping.parent.authority_provided_id
+                )
```

Comment:
Find the `authority_provided_id` of the group's course. (Probably worth a comment.)